# Лабораторная работа 2-1

Сергеева Анастасия АДЭУ-211

## 2.1.1: Множества

Напишите функцию, которая находит пересечение двух множеств и печатает их.
Инициализируйте два множества, например, значениями {13, 25, 37, 45, 13} и {14, 25, 38, 8, 45} и вызовите  функцию.

In [ ]:
#созданы 2 множества
mn1 = {13, 25, 37, 45, 13}
mn2 = {14, 25, 38, 8, 45}

#поиск пересечения 2 множеств
print(mn1 & mn2)

{25, 45}


## 2.1.2: Dictionaries

* Создайте словарь с двухбуквенными кодами двух федеральных округов России и полными названиями, например, ЦТ: 	Центральный, ЮЖ: Южный.
* После первоначального создания словаря добавьте в словарь еще два округа.
* Создайте второй словарь, который сопоставляет коды федеральных округов России с массивом городов в этом округе, например, ЦТ: [Москва, Тверь, Химки, Мытищи].
* Напишите функцию, которая принимает код федеральных округов, печатает полное название округа и выводит список городов в этом округе.

Подготовлены 2 словаря

In [ ]:
#создан словарь
sl1 = {"ЦТ":"Центральный", "ЮЖ":"Южный"}

#добавлены значения в этот словарь
sl1["СЗ"] = "Северо-Западный"
sl1["СК"] = "Северо-Кавказский"

#создан 2 словарь
sl2 = {"ЦТ":["Москва", "Смоленск"], "ЮЖ":["Адлер", "Адыгея","Азов"], "СЗ":["Архангельск", "Северодвинск"], "СК":["Махачкала", "Ставрополь"]}

Напишите функцию, которая принимает код федеральных округов, печатает полное название округа и выводит список городов в этом округе.

In [ ]:
#создана функция, выводящая коды федеральных округов, их название и города
def f1():
  for k in sl1.keys():
    okr = sl1[k]
    city = sl2[k]
    print(f"Код ФО: {k}, Название ФО: {okr}, Города: {city}")
f1()

Код ФО: ЦТ, Название ФО: Центральный, Города: ['Москва', 'Смоленск']
Код ФО: ЮЖ, Название ФО: Южный, Города: ['Адлер', 'Адыгея', 'Азов']
Код ФО: СЗ, Название ФО: Северо-Западный, Города: ['Архангельск', 'Северодвинск']
Код ФО: СК, Название ФО: Северо-Кавказский, Города: ['Махачкала', 'Ставрополь']


## 2.1.3: Objects

Создайте класс `Pet` с параметрами `name, pronoun, animal,` и `pet_response`.

Добавьте метод `pet()`, который печатает ответ, состоящий из элементов, например:

` Лейла — кошка. Если ее погладить, Лейла мурчит».`
` Скаут — собака. Если его погладить, Скаут виляет хвостом.`

In [ ]:
class Pet:
    # создание экземпляра
    def __init__(a, name, pronoun, animal, pet_response):
        # переменная экземпляра
        a.name = name
        a.pronoun = pronoun
        a.animal = animal
        a.pet_response = pet_response

    # метод, печатающий ответ
    def print_Pet(a):
        print(a.name, " - ", a.animal, ", If you stroke" , a.pronoun, ", ", a.name, a.pet_response, ".", sep="")

Вот несколько примеров вызовов:

In [ ]:
layla = Pet("Layla", "her", "cat", "purrs")
scout = Pet("Scout", "her", "dog", "wags her tail")
layla.print_Pet()
scout.print_Pet()

Layla - cat, If you strokeher, Laylapurrs.
Scout - dog, If you strokeher, Scoutwags her tail.


## 2.1.4: Pandas Series

Создайте `pandas series`, используя приведенные ниже списки, содержащие названия команд [Российской премьер-лиги](https://ru.wikipedia.org/wiki/%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B9%D1%81%D0%BA%D0%B0%D1%8F_%D0%BF%D1%80%D0%B5%D0%BC%D1%8C%D0%B5%D1%80-%D0%BB%D0%B8%D0%B3%D0%B0) и количество выигранных ими кубка премьер-лиги. Используйте имена как индексы, а победы как данные.

  * После создания списка отсортируйте данные в алфавитном порядке по индексу.
  * Распечатать обзор статистических свойств ряда. Каково среднее количество побед?
  * Отфильтровать все команды, выигравшие менее четырех кубков премьер-лиги.
  * В футбольной команде 11 игроков. Обновите данные, чтобы вместо количества чемпионатов,  отражалось количество побед.
  * Предположим, что каждая  победа приносила 50 000 долларов США. Обновите данные, чтобы они содержали строку суммы в долларах, включая знак \$. Например, для `Локомотив` это будет соответствовать:
  ```
  Тренер Юрий Сёмин  Локомотив $ 600000
  ```

Выгрузка команд и количества их побед из Википедии

In [2]:
#импортируем библиотеки
import requests
import pandas as pd

#импортируем в датафрейм
url="https://ru.wikipedia.org/wiki/Российская_премьер-лига#Тренеры"
req=requests.get(url)
data=pd.read_html(req.text)
df=data[5]

# Создаем pandas series
ser = df.set_index('Тренер')['Чемпионат']
print(ser)

Тренер
Юрий Сёмин («Локомотив»)                    3
Олег Романцев («Спартак»)                   8
Сергей Семак («Зенит»)                      5
Валерий Газзаев («Спартак-Алания», ЦСКА)    4
Леонид Слуцкий (ЦСКА)                       3
Курбан Бердыев («Рубин»)                    2
Лучано Спаллетти («Зенит»)                  2
Андре Виллаш-Боаш («Зенит»)                 1
Дик Адвокат («Зенит»)                       1
Массимо Каррера («Спартак»)                 1
Зико (ЦСКА)                                 0
Георгий Ярцев («Спартак»)                   1
Юрий Миронов («Торпедо»)                    0
Константин Бесков («Динамо»)                0
Анатолий Давыдов («Зенит»)                  0
Ваит Талгаев («Терек»)                      0
Анатолий Бышовец («Локомотив»)              0
Миодраг Божович («Ростов»)                  0
Игорь Черевченко («Локомотив»)              0
Дмитрий Парфёнов («Тосно»)                  0
Марко Николич («Локомотив»)                 0
Паоло Ваноли («Спартак»)   

Сортировка по индексу в алфавитном порядке

In [3]:
ser.sort_index(ascending=True)

Тренер
Анатолий Бышовец («Локомотив»)              0
Анатолий Давыдов («Зенит»)                  0
Андре Виллаш-Боаш («Зенит»)                 1
Артур Жорже (ЦСКА)                          0
Ваит Талгаев («Терек»)                      0
Валерий Газзаев («Спартак-Алания», ЦСКА)    4
Виктор Гончаренко (ЦСКА)                    0
Владимир Федотов (ЦСКА)                     0
Властимил Петржела («Зенит»)                0
Георгий Ярцев («Спартак»)                   1
Дик Адвокат («Зенит»)                       1
Дмитрий Парфёнов («Тосно»)                  0
Зико (ЦСКА)                                 0
Игорь Черевченко («Локомотив»)              0
Константин Бесков («Динамо»)                0
Курбан Бердыев («Рубин»)                    2
Леонид Слуцкий (ЦСКА)                       3
Лучано Спаллетти («Зенит»)                  2
Марко Николич («Локомотив»)                 0
Массимо Каррера («Спартак»)                 1
Миодраг Божович («Ростов»)                  0
Мирча Луческу («Зенит»)    

Обзор статистических свойств ряда.

In [4]:
ser.describe()

count    27.000000
mean      1.148148
std       1.955342
min       0.000000
25%       0.000000
50%       0.000000
75%       1.500000
max       8.000000
Name: Чемпионат, dtype: float64

Всего побед 27, среднее количество побед -  1.148148, максимум 8 побед, минимум ноль.

Отфильтровать все команды, выигравшие менее четырех кубков премьер-лиги.

In [5]:
#Создаем логическую маску для фильтрации
mask = ser < 4
mask

Тренер
Юрий Сёмин («Локомотив»)                     True
Олег Романцев («Спартак»)                   False
Сергей Семак («Зенит»)                      False
Валерий Газзаев («Спартак-Алания», ЦСКА)    False
Леонид Слуцкий (ЦСКА)                        True
Курбан Бердыев («Рубин»)                     True
Лучано Спаллетти («Зенит»)                   True
Андре Виллаш-Боаш («Зенит»)                  True
Дик Адвокат («Зенит»)                        True
Массимо Каррера («Спартак»)                  True
Зико (ЦСКА)                                  True
Георгий Ярцев («Спартак»)                    True
Юрий Миронов («Торпедо»)                     True
Константин Бесков («Динамо»)                 True
Анатолий Давыдов («Зенит»)                   True
Ваит Талгаев («Терек»)                       True
Анатолий Бышовец («Локомотив»)               True
Миодраг Божович («Ростов»)                   True
Игорь Черевченко («Локомотив»)               True
Дмитрий Парфёнов («Тосно»)                 

In [6]:
# применение маски
ser1=ser[mask]

#сбрасываем индексацию
ser1.reset_index()

Тренер  Чемпионат
0         Юрий Сёмин («Локомотив»)          3
1            Леонид Слуцкий (ЦСКА)          3
2         Курбан Бердыев («Рубин»)          2
3       Лучано Спаллетти («Зенит»)          2
4      Андре Виллаш-Боаш («Зенит»)          1
5            Дик Адвокат («Зенит»)          1
6      Массимо Каррера («Спартак»)          1
7                      Зико (ЦСКА)          0
8        Георгий Ярцев («Спартак»)          1
9         Юрий Миронов («Торпедо»)          0
10    Константин Бесков («Динамо»)          0
11      Анатолий Давыдов («Зенит»)          0
12          Ваит Талгаев («Терек»)          0
13  Анатолий Бышовец («Локомотив»)          0
14      Миодраг Божович («Ростов»)          0
15  Игорь Черевченко («Локомотив»)          0
16      Дмитрий Парфёнов («Тосно»)          0
17     Марко Николич («Локомотив»)          0
18        Паоло Ваноли («Спартак»)          0
19         Владимир Федотов (ЦСКА)          0
20    Властимил Петржела («Зенит»)          0
21              Артур Жорже (ЦСКА)          0
22         Мирча Луческу («Зенит»)          0
23        Виктор Гончаренко (ЦСКА)          0

В футбольной команде 11 игроков. Обновите данные, чтобы вместо количества чемпионатов, отражалось количество побед.

In [7]:
#переименовываем столбец Чемпионат на Победы
ser2= ser.rename("Победы")

#сбрасываем индексацию
ser2.reset_index()

Тренер  Победы
0                   Юрий Сёмин («Локомотив»)       3
1                  Олег Романцев («Спартак»)       8
2                     Сергей Семак («Зенит»)       5
3   Валерий Газзаев («Спартак-Алания», ЦСКА)       4
4                      Леонид Слуцкий (ЦСКА)       3
5                   Курбан Бердыев («Рубин»)       2
6                 Лучано Спаллетти («Зенит»)       2
7                Андре Виллаш-Боаш («Зенит»)       1
8                      Дик Адвокат («Зенит»)       1
9                Массимо Каррера («Спартак»)       1
10                               Зико (ЦСКА)       0
11                 Георгий Ярцев («Спартак»)       1
12                  Юрий Миронов («Торпедо»)       0
13              Константин Бесков («Динамо»)       0
14                Анатолий Давыдов («Зенит»)       0
15                    Ваит Талгаев («Терек»)       0
16            Анатолий Бышовец («Локомотив»)       0
17                Миодраг Божович («Ростов»)       0
18            Игорь Черевченко («Локомотив»)       0
19                Дмитрий Парфёнов («Тосно»)       0
20               Марко Николич («Локомотив»)       0
21                  Паоло Ваноли («Спартак»)       0
22                   Владимир Федотов (ЦСКА)       0
23              Властимил Петржела («Зенит»)       0
24                        Артур Жорже (ЦСКА)       0
25                   Мирча Луческу («Зенит»)       0
26                  Виктор Гончаренко (ЦСКА)       0

Предположим, что каждая победа приносила 50 000 долларов США. Обновите данные, чтобы они содержали строку суммы в долларах, включая знак $. Например, для Локомотив это будет соответствовать:

In [8]:
#умножаем на доллары
ser3 = ser * 50000
ser3

Тренер
Юрий Сёмин («Локомотив»)                    150000
Олег Романцев («Спартак»)                   400000
Сергей Семак («Зенит»)                      250000
Валерий Газзаев («Спартак-Алания», ЦСКА)    200000
Леонид Слуцкий (ЦСКА)                       150000
Курбан Бердыев («Рубин»)                    100000
Лучано Спаллетти («Зенит»)                  100000
Андре Виллаш-Боаш («Зенит»)                  50000
Дик Адвокат («Зенит»)                        50000
Массимо Каррера («Спартак»)                  50000
Зико (ЦСКА)                                      0
Георгий Ярцев («Спартак»)                    50000
Юрий Миронов («Торпедо»)                         0
Константин Бесков («Динамо»)                     0
Анатолий Давыдов («Зенит»)                       0
Ваит Талгаев («Терек»)                           0
Анатолий Бышовец («Локомотив»)                   0
Миодраг Божович («Ростов»)                       0
Игорь Черевченко («Локомотив»)                   0
Дмитрий Парфёнов («Тосно

In [11]:
#создаем метод для приписания знака доллара
def toDollars(amount):
    return "$ " + str(amount)

#применяем метод
ser3.map(toDollars).reset_index()

Тренер Чемпионат
0                   Юрий Сёмин («Локомотив»)  $ 150000
1                  Олег Романцев («Спартак»)  $ 400000
2                     Сергей Семак («Зенит»)  $ 250000
3   Валерий Газзаев («Спартак-Алания», ЦСКА)  $ 200000
4                      Леонид Слуцкий (ЦСКА)  $ 150000
5                   Курбан Бердыев («Рубин»)  $ 100000
6                 Лучано Спаллетти («Зенит»)  $ 100000
7                Андре Виллаш-Боаш («Зенит»)   $ 50000
8                      Дик Адвокат («Зенит»)   $ 50000
9                Массимо Каррера («Спартак»)   $ 50000
10                               Зико (ЦСКА)       $ 0
11                 Георгий Ярцев («Спартак»)   $ 50000
12                  Юрий Миронов («Торпедо»)       $ 0
13              Константин Бесков («Динамо»)       $ 0
14                Анатолий Давыдов («Зенит»)       $ 0
15                    Ваит Талгаев («Терек»)       $ 0
16            Анатолий Бышовец («Локомотив»)       $ 0
17                Миодраг Божович («Ростов»)       $ 0
18            Игорь Черевченко («Локомотив»)       $ 0
19                Дмитрий Парфёнов («Тосно»)       $ 0
20               Марко Николич («Локомотив»)       $ 0
21                  Паоло Ваноли («Спартак»)       $ 0
22                   Владимир Федотов (ЦСКА)       $ 0
23              Властимил Петржела («Зенит»)       $ 0
24                        Артур Жорже (ЦСКА)       $ 0
25                   Мирча Луческу («Зенит»)       $ 0
26                  Виктор Гончаренко (ЦСКА)       $ 0